tests/unit/orchestrate.py

```diff
@@ -0,0 +1,356 @@
+    because this is what was run to produce the counts and the initial test sequence.
+    Now `orchestrate` will run all of the remaining scenarios by finding all the
+    permutations of the test sequence and executing those, in turn::
+
+        [0, 1, 0, 1]
+        [0, 1, 1, 0]
+        [1, 0, 0, 1]
+        [1, 0, 1, 0]
+        [1, 1, 0, 0]
+
+    You'll notice in our example that since both test functions are actually the same
+    function, that although it tested 6 scenarios there are effectively only really 3
+    unique scenarios. For the time being, though, `orchestrate` doesn't attempt to
+    detect this condition or optimize for it.
+
+    There are some performance considerations that should be taken into account when
+    writing tests. The number of unique test sequences grows quite quickly with the
+    number of turns taken by the functions under test. Our simple example with two
+    threads each taking two turns, only yielded 6 scenarios, but two threads each taking
+    6 turns, for example, yields 924 scenarios. Add another six step thread and now you
+    have over 17 thousand scenarios. In general, use the least number of steps/threads
+    you can get away with and still expose the behavior you want to correct.
+
+    For the same reason as above, if you have many concurrent tests, when writing a new
+    test, make sure you're not accidentally patching syncpoints intended for other
+    tests, as this will add steps to your tests. While it's not problematic from a
+    testing standpoint to have extra steps in your tests, it can use computing resources
+    unnecessarily. Using different no-op functions with different names for different
+    tests can help with this.
+
+    As soon as any error or failure is detected, no more scenarios are run
+    and that error is propagated to the main thread.
+
+    Args:
+        tests (Tuple[Callable]): Test functions to be run. These functions will not be
+            called with any arguments, so they must not have any required arguments.
+
+    Returns:
+        Tuple[int]: A tuple of the count of the number turns for test passed in. Can be
+            used a sanity check in tests to make sure you understand what's actually
+            happening during a test.
+    """
+    # Produce an initial test sequence. The fundamental question we're always trying to
+    # answer is "whose turn is it?" First we'll find out how many "turns" each test
+    # needs to complete when run serially and use that to construct a sequence of
+    # indexes. When a test's index appears in the sequence, it is that test's turn to
+    # run. We'll start by constructing a sequence that would run each test through to
+    # completion serially, one after the other.
+    test_sequence = []
+    counts = []
+    for index, test in enumerate(tests):
+        thread = _TestThread(test)
+        for count in itertools.count(1):  # pragma: NO BRANCH
```

Comment:
The examples in the docstring above seem to use `0` as the starting index into the set of permuted indexs:  this makes `1` the lowest index, right?

Reply:
0 is the starting index, as held in the `index` two lines up. `count` is the number of "turns" a test takes to complete.